README.md
python-corrida
==============

A Python translation of `corrida.for`.

The original FORTRAN program `corrida.for` was written by Paulo Polito to
create a training spreadsheet for running.


CHANGES.txt
2013-08-07 -- First python version.


corrida/corrida.for
      program corrida
      implicit none
      real*8 x(27),pvm(10),phr(10),y,t,d,p,s,v,vm
      real*8 vo2,pvmax,vo2max,phrmax,hrmax,hr
      real*8 estd,estt,t1(10,27),r1(10,27)
      integer*4 hh1(10,27),mm1(10,27),ss1(10,27)
      integer*4 rh1(10,27),rm1(10,27),rs1(10,27)
      integer*4 hh,mm,ss
      integer*4 i,j
      character*1 co,ps,sc
      character*10 name
      write(*,*)'Entre o seu nome'
      read(*,'(a10)')name
      write(*,*)'Entre a sua idade'
      read(*,*)y
      write(*,*)'Entre com a distancia base em m'
      read(*,*)d
      write(*,*)'Entre com o tempo HH:MM:SS'
      read(*,'(i2,1x,i2,1x,i2)')hh,mm,ss

      t=float(hh)*60.+float(mm)+float(ss)/60
      co=':'
      sc=','
      ps='%'
      
      do i=1,10
         pvm(i)=0.55+0.05*(float(i))
         phr(i)=0.63*pvm(i)+0.37
      end do

      x(1)=100.
      x(2)=120.
      x(3)=150.
      x(4)=200.
      x(5)=300.
      x(6)=400.
      x(7)=500.
      x(8)=600.
      x(9)=667.
      x(10)=800.
      x(11)=1000.
      x(12)=1200.
      x(13)=1500.
      x(14)=1600.
      x(15)=2000.
      x(16)=3000.
      x(17)=5000.
      x(18)=6000.
      x(19)=7000.
      x(20)=8000.
      x(21)=10000.
      x(22)=12000.
      x(23)=15000.
      x(24)=18000.
      x(25)=20000.
      x(26)=42195./2.
      x(27)=42195.

      v=vo2(t,d)
      p=pvmax(t)
      vm=vo2max(v,p)

      hrmax= 206.3 - (0.711 * y)
      phrmax=.63*p+.37
      hr=phrmax*hrmax
      do j=1,27
         do i=10,1,-1
            t1(i,j)=60*estt(x(j),pvm(i)*vm)
            call d2hms(t1(i,j),hh1(i,j),mm1(i,j),ss1(i,j))
            r1(i,j)=1000.*t1(i,j)/x(j)            
            call d2hms(r1(i,j),rh1(i,j),rm1(i,j),rs1(i,j))
         end do
      end do
      open(9,file='corrida.csv',status='unknown')
      write(9,'(a9,a10)')'Atleta:, ',name
      write(9,'(a9,i2,a6)')'Idade:, ',int(y),',anos'
      write(9,'(a17,f8.1,a2)')'Distancia,base:,',d,',m'
      write(9,'(a13,i2.2,1a,i2.2,1a,i2.2)')'Tempo,base:,',hh,co,mm,co,ss
      write(9,*)' '
      write(9,'(a15,i3,a10)')'VO2,estimado:,',Nint(v),',ml/kg/min'
      write(9,'(a15,i3,a18,i3,a10)')'Equivalente,a,',Nint(p*100),
     $     '%,do,VO2,maximo:,',Nint(vm),',ml/kg/min'
      write(9,*)' '
      write(9,'(a17,i3,a4)')'Ritmo,cardiaco:,', Nint(hr),',bpm'
      write(9,'(a15,i3,a28,i3,a4)')'Equivalente,a,',Nint(phrmax*100),
     $     '%,do,ritmo,cardiaco,maximo:,',Nint(hrmax),',bpm'
      write(9,*)' '
      write(9,*)'Tempos,Estimados,e,%VO2,Max '
      write(9,'(a9,10(1a,4x,i3,1a))')'Distancia',
     $     (sc,Nint(pvm(i)*100),ps,i=10,1,-1)
      write(9,*)' '
      write(9,'(27(f8.1,10(1a,i2.2,1a,i2.2,1a,i2.2),/))')
     $ (x(j),(sc,hh1(i,j),co,mm1(i,j),co,ss1(i,j),i=10,1,-1),j=1,27)
      write(9,*)'Ritmos,Estimados,e,%Max,Ritmo,Cardiaco '
      write(9,'(a9,10(1a,4x,i3,1a))')'Distancia',
     $     (sc,Nint(phr(i)*100),ps,i=10,1,-1)
      write(9,*)' '
      write(9,'(27(f8.1,10(1a,i2.2,1a,i2.2,1a,i2.2),/))')
     $ (x(j),(sc,rh1(i,j),co,rm1(i,j),co,rs1(i,j),i=10,1,-1),j=1,27)
      write(9,*)' '
      close(9)

      end

      subroutine d2hms(t,hh,mm,ss)
      implicit none
      real*8 t
      integer*4 hh,mm,ss
      hh=int(t/3600.)
      mm=int(t/60-float(hh*60) )
      ss=Nint(t-float(mm*60)-float(hh*3600))
      return
      end

      function vo2(t,d)
      implicit none
      real*8 t,d,s,a,b,r,vo2
      a=-4.60
      b=0.182258
      r=0.000104
      s=d/t
      vo2=(a+b*s+r*s**2)
      return
      end

      function pvmax(t)
      implicit none
      real*8 d,e,f,g,h,t,pvmax
      d=0.8
      e=0.1894393
      f=-0.012778
      g=0.2989558
      h=-0.1932605
      pvmax=(d+e*exp(f*t)+g*exp(h*t))
      return
      end

      function vo2max(v,p)
      implicit none
      real*8 v,p,vo2max
      vo2max=v/p
      return
      end

      function estt(estd,vm)
      implicit none
      real*8 vm,vo2,pvmax,vo2max,eps,aeps,t1,v1,p1,vm1,dt
      real*8 estd,estt
      estt=estd/50
      dt=estt/2
      aeps=9999
      do while(aeps.gt.1e-6)
         v1=vo2(estt,estd)
         p1=pvmax(estt)
         vm1=vo2max(v1,p1)
         eps=vm1-vm
         aeps=abs(eps)
         if(eps.gt.1e-6)then ! increase estt
            estt=estt+dt
         elseif(eps.lt.-1e-6)then ! decrese estt
            estt=estt-dt
         end if
         dt=dt/2
      end do
      return
      end



corrida/corrida.py
# -*- coding: utf-8 -*-
#
# corrida.py
#
# purpose:  Reproduce "corrida.for" results.
# author:   Filipe P. A. Fernandes
# e-mail:   ocefpaf@gmail
# web:      http://ocefpaf.tiddlyspot.com/
# created:  27-May-2013
# modified: Fri 07 Jun 2013 06:38:49 PM BRT
#
# obs:
#

from __future__ import division

import numpy as np
from pandas import DataFrame
from cStringIO import StringIO

# Constants.
encoding = 'UTF-8'
pvm = 0.55 + 0.05 * (np.arange(10) + 1)
phr = 0.63 * pvm + 0.37

x = np.array([100, 120, 150, 200, 300, 400, 500, 600, 667, 800, 1000, 1200,
              1500, 1600, 2000, 3000, 5000, 6000, 7000, 8000, 10000, 12000,
              15000, 18000, 20000, 42195 / 2, 42195])


# Utilities.
def Nint(num):
    return np.int_(np.round(num))


def to_csv(df):
    df.index.name = u'Distância'
    csv = StringIO()
    df.to_csv(csv, encoding=encoding)
    csv.seek(0)
    return [line.strip().decode(encoding=encoding) for line in csv.readlines()]


def write_spreadsheet(fname='corrida.csv', lines=None):
    with open(fname, 'w') as f:
        f.writelines(u'\n'.join(lines).encode(encoding))
    return None


# Functions.
def d2hms(t):
    hh = t // 3600
    mm = t // 60 - hh * 60
    ss = Nint(t - mm * 60 - hh * 3600)
    return "%02i:%02i:%02i" % (hh, mm, ss)


def vo2(t, d):
    a, b, r = -4.60, 0.182258, 0.000104
    s = d / t
    return a + b * s + r * s ** 2


def pvmax(t):
    d, e, f, g, h = 0.8, 0.1894393, -0.012778,  0.2989558, -0.1932605
    return d + e * np.exp(f * t) + g * np.exp(h * t)


def vo2max(v, p):
    return v / p


def _estt(estd, vm):
    estd, vm = np.broadcast_arrays(estd[..., None], vm[None, ...])
    estt = estd / 50
    dt = estt / 2
    aeps = 9999
    while aeps > 1e-6:
        v1, p1 = vo2(estt, estd),  pvmax(estt)
        vm1 = vo2max(v1, p1)
        eps = vm1 - vm
        aeps = np.abs(eps).min()
        mask = eps > 1e-6  # Increase estt.
        estt[mask] += dt[mask]
        mask = eps < -1e-6  # Decrease estt.
        estt[mask] -= dt[mask]
        dt /= 2
    return estt

if __name__ == '__main__':
    if False:
        hh, mm, ss = 00, 23, 00
        name, y, d = 'Filipe', 33, 5000

    if True:
        name = raw_input('Name: ')
        y = int(raw_input('Age: '))
        d = float(raw_input('Distance in meters: '))
        time = raw_input('Time HH MM SS: ')
        hh, mm, ss = [int(n) for n in time.split()]

    t = float(hh) * 60. + float(mm) + float(ss) / 60
    v = vo2(t, d)
    p = pvmax(t)
    vm = vo2max(v, p)
    hrmax = 206.3 - (0.711 * y)
    phrmax = 0.63 * p + 0.37
    hr = phrmax * hrmax

    t1 = np.fliplr(60 * _estt(x, pvm * vm))
    r1 = np.fliplr(1000 * t1 / x[..., None])

    cols0 = ('%i%%,%i%%,%i%%,%i%%,%i%%,%i%%,%i%%,%i%%,%i%%,%i%%' %
             tuple(Nint(pvm * 100))[::-1])
    cols1 = ('%i%%,%i%%,%i%%,%i%%,%i%%,%i%%,%i%%,%i%%,%i%%,%i%%' %
             tuple(Nint(phr * 100))[::-1])

    t1 = DataFrame(t1, index=x, columns=cols0.split(',')).applymap(d2hms)
    r1 = DataFrame(r1, index=x, columns=cols1.split(',')).applymap(d2hms)

    # Text output.
    lines0 = [
        u'Atleta:,%s' % name,
        u'Idade:,%i,anos' % y,
        u'Distância,base:,%s,m' % d,
        u'Tempo,base:,%02i:%02i:%02i' % (hh, mm, ss),
        u'',
        u'VO2,estimado:,%i,ml/kg/min' % np.round(v),
        (u'Equivalente,a,%i%%,do,VO2,máximo:%i,ml/kg/min' %
         (Nint(p * 100), Nint(vm))),
        u'',
        u'Ritmo,cardíaco:,%i,bpm' % Nint(hr),
        (u'Equivalente,a,%i%%,do,ritmo,cardíaco,máximo:,%i,bpm' %
         (Nint(phrmax * 100), Nint(hrmax))),
        u'',
        u'Tempos,Estimados,e,%VO2,Max',
        u'']
    lines1 = to_csv(t1)
    lines2 = [u'', u'Ritmos,Estimados,e,%Max,Ritmo,Cardíaco']
    lines3 = to_csv(r1)

    write_spreadsheet('corrida.csv', lines0 + lines1 + lines2 + lines3)
